File: mnist_imp/python/activation.py
import numpy as np

def ReLu(X):
    return np.maximum(0, X)

def softmax(X, clip=True):
    maxa = np.max(X, axis=1, keepdims=True)
    X = X - maxa
    probs = np.exp(X)
    probs = probs / np.sum(probs, axis=1, keepdims=True)
    if clip:
        probs = np.clip(probs, 1e-10, 1.0)
    return probs

File: mnist_imp/cpp/src/activation.cpp
//
// Created by Prateek Bansal on 8/1/25.
//
#include "Eigen/Dense"

Eigen::ArrayXXf ReLu(const Eigen::ArrayXXf& input)
{
    return input.max(0.0f);
}

Eigen::ArrayXXf softmax(const Eigen::ArrayXXf& X, bool clip = true) {
    Eigen::ArrayXXf X_shifted = X.colwise() - X.rowwise().maxCoeff();
    Eigen::ArrayXXf exps = X_shifted.exp();
    Eigen::ArrayXXf probs = exps.colwise() / exps.rowwise().sum();

    if (clip) {
        probs = probs.max(1e-10f).min(1.0f);
    }
    return probs;
}